keps/sig-node/1898-hardened-exec/README.md

```diff
@@ -0,0 +1,537 @@
+
+- If only body options are provided, use those.
+- If only query options are provided, use those.
+- If both body and query options are provided, require that they be identical.
+
+Eventually, body options can be required. The requirement will be controlled by the
+`HardenedExecRequests` feature gate. The feature will remain in alpha for TBD releases after
+supported clients have been updated with the new request format.
+
+<<[UNRESOLVED]>>
+
+OPEN QUESTION: How many releases should we wait before requiring request body parameters?
+
+<<[/UNRESOLVED]>>
+
+To minimize risk of breakage, attach and port-forward requests should be left unchanged, as they do
+not share the same risks of arbitrary code execution as exec.
+
+#### 2. Match new Kubelet request requirements.
+
+The API server must be updated to unconditionally send POST requests to the Kubelet (for
+exec/attach/port-forward), and provide the request options in the body. See [2. Require a
+<code>POST</code> request to the streaming
+endpoints](#2-require-a--request-to-the-streaming-endpoints) and [4. Require the request options to
+be provided in the request body](#4-require-the-request-options-to-be-provided-in-the-request-body).
+
+#### 3. Require the websocket protocol for GET requests.
+
+We cannot completely remove support for GET exec requests without breaking websockets. However, we
+can require the websocket protocol be used to perform a GET exec.
+
+**Risk:** This is a breaking change for non-websocket clients, and leaves websocket clients exposed
+to SSRF risks.
+
+<<[UNRESOLVED]>>
+
+OPEN QUESTION: How common is it for non-websocket clients to use GET for exec requests? `kubectl`
+does not. Do we need to do a gradual rollout of this change?
+
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED]>>
+
+Alternative options for protecting websockets at the expense of introducing a breaking change:
+
+1. As extra protection against redirect attacks, we could use a similar trick as is used for
+providing the bearer token as an extra sub-protocol:
+https://github.com/kubernetes/kubernetes/blob/21953d15ea48972f20a8de29d58bd5ce6d913914/staging/src/k8s.io/apiserver/pkg/authentication/request/websocket/protocol.go#L37-L38
+In this case, we could take a hash of the request options and provide the hash as an additional
+sub-protocol. Then, the exec handler would need to validate that the request options matched the
+hash value in the sub-protocol.
+
+2. Approve the exec params with a separate POST request. For example, the client issues a POST to
+   `.../exec/token` with the exec params included in the request body. The response includes a token
+   that must be included in the GET exec params. Implementation options include caching a single-use
+   token (similar to how CRI streaming requests work), or signing the request params.
+
+3. The initial websocket request only opens the websocket protocol stream, and a subsequent request
+   must be sent over the websocket to initiate the actual GET request.
```

Comment:
```suggestion
    must be sent over the websocket to initiate the actual exec or other action.
```
(because WebSocket, as a byte stream, doesn't have HTTP verbs)